pandas/_libs/groupby_helper.pxi.in

```diff
@@ -618,6 +618,75 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
             for i in range(N):
                 out[i, 0] = out[i, 0] / grp_sizes[i, 0]
 {{endif}}
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def group_fillna_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
+                          ndarray[{{c_type}}, ndim=2] values,
+                          ndarray[int64_t] labels,
+                          object method,
+                          int64_t limit):
+    """Fills values forwards or backwards within a group
+
+    Parameters
+    ----------
+    out : array of {{dest_type2}} values which this method will write its
+        results to
+    values : array of {{c_type}} values which may require filling
+    labels : array containing unique label for each group, with its ordering
+        matching up to the corresponding record in `values`
+    method : {'ffill', 'bfill'}
+        Direction for fill to be applied (forwards or backwards, respectively)
+    limit : Consecutive values to fill before stopping, or -1 for no limit
+
+    Notes
+    -----
+    This method modifies the `out` parameter rather than returning an object
+    """
+    cdef:
+        Py_ssize_t i, N
+        ndarray[uint8_t] mask
+        ndarray[int64_t] sorted_labels
+        {{dest_type2}} curr_fill_val = {{nan_val}}
+        int64_t idx, filled_vals=0
+
+    N, K = (<object> values).shape
+
+    {{if name=='int64'}}
+    mask = (values[:, 0] == {{nan_val}}).astype(np.uint8)
+    {{elif name=='object'}}
+    mask = np.array([x != x for x in values[:, 0]]).astype(np.uint8)
+    {{else}}
+    mask = np.isnan(values[:, 0]).astype(np.uint8)
+    {{endif}}
+
+    sorted_labels = np.argsort(labels)
+    if method == 'bfill':
+        sorted_labels[::-1].sort()
```

Comment:
Why are the labels reversed before sorting?

Reply:
They only get reversed in the case of `bfill` - it's the cheapest way to fill backwards that I could think of

Reply:
Right, but doesn't the `.sort()` undo the reversing?

Reply:
I took that from the below SO article and the comment by perimosocordiae was helpful in figuring out what's going on. Open to changing it to `sorted_labels = np.sort(sorted_labels)[::-1]` for readability if you'd like

https://stackoverflow.com/questions/26984414/efficiently-sorting-a-numpy-array-in-descending-order

Reply:
Oh, thanks! Perhaps too clever, how about this?  still avoids the copy
```
sorted_labels.sort()
sorted_labels = sorted_labels[::-1]
```

Reply:
Actually I think you are on to something - the extra `sort` doesn't feel right and I think it could cause a bug as it will really just return an array ranging from N-1..0. Going to work up some test cases to validate and add fix to next commit